proposals/0000-forall-arrow.rst

```diff
@@ -0,0 +1,158 @@
+                     accepted.
+
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/81>`_. 
+
+.. contents::
+
+A syntax for visible dependent quantification
+=============================================
+
+GHC 8.0 has support for visible dependent quantification in kinds. For example, GHC 8 will accept ::
+
+  data T k (a :: k)
+
+Note that any (fully-applied) use of ``T`` has to mention both a kind and a type. For example,
+``T Nat 3`` and ``T Type Int`` are well-typed. If you ask GHCi for the kind of ``T``, it tells
+you ``forall k -> k -> *`` (with ``-fprint-explicit-foralls``). Note that there is no ``.`` after
+the ``forall``. Instead, this kind makes ``k`` a visible, dependent type variable. It's *visible*
+because every (fully-applied) use of ``T`` must pass in a value for ``k`` visibly (explicitly).
+It's *dependent* because ``k`` appears later in the kind of ``T``. The ability to have visible,
+dependent quantifiees is a new feature in GHC 8, available only in kinds. (There's no way to do
+this in terms.)
+
+The syntax ``forall k -> k -> *`` that is printed in GHCi for these kinds is not currently parsed.
+This proposal proposed to add this kind format to the language as a first-class kind.
+
+This proposal is viable on its own, but it may be best considered in the context of other
+quantifiers needed to support dependent types. See `#102 <https://github.com/goldfirere/ghc-proposals/blob/pi/proposals/0000-pi.rst>`_.
+
+Motivation
+------------
+
+Currently, GHC allows the user to define a type with a visible, dependent quantifiers in its kind,
+but it offers no way to write the kind directly. This is just plain silly.
+
+(Historical note: the reason this happened is that I got early feedback claiming that ``forall k -> ...``
+was poor syntax, and I was reticent to add it to the language with ``-XTypeInType``. However, now that
+this proposals process is active, I'm happy to propose the new syntax where it can get debated in the
+open and perhaps refined.)
+
+A useful example of how this construct might be used is an alternative syntax for ``TypeRep``. We can
+imagine ::
+
+  TypeRepV :: forall k -> k -> Type   -- "V" for visible
+
+as a companion to ::
+
+  TypeRep :: forall k. k -> Type      -- we have this today
+
+Now, if someone wants to operate on
+a type representation for some type of kind ``Type -> Type``, they could say ::
+
+  foo :: TypeRepV (Type -> Type) a -> ...
+
+and GHC would easily infer that ``a`` should have kind ``Type -> Type``.  
```

Comment:
Add something like:

In contrast, with `TypeRep` where `foo` would be written as
```haskell
foo :: TypeRep a -> …
```